File: week1-array-string-hashing/contain_duplicate.cpp
#include "unordered_set"
#include "vector"
using std::unordered_set;
using std::vector;

bool contain_duplicate(vector<int>& nums) {
    unordered_set<int> seen;
    for (int n : nums) {
        if (!seen.insert(n).second) return true;
    }
    return false;
}

File: week1-array-string-hashing/contain_duplicate.rs
use std::collections::HashSet;

fn contain_duplicate(nums: Vec<i32>) -> bool {
    let mut seen: HashSet<i32> = HashSet::new();
    for &num in nums.iter() {
        if seen.contains(&num) {
            return true;
        }
        seen.insert(num);
    }
    false
}

fn contain_duplicate_iterator(nums: Vec<i32>) -> bool {
    let mut seen = HashSet::new();
    nums.iter().any(|&num| !seen.insert(num))
}

fn main() {
    println!("{:?}", contain_duplicate(vec![2, 7, 11, 15]));
    println!("{:?}", contain_duplicate(vec![3, 2, 2, 4]));
    println!("{:?}", contain_duplicate(vec![3, 3]));
}
